File: frontend/static/js/payroll.js
/* Payroll & Accounting — Super Admin ONLY.
   Reads /api/payroll?period=YYYY-MM (net pay computed from attendance + overtime + leave), lets the
   Super Admin set each person's monthly salary and a per-period bonus/deduction, finalize a month,
   and export the sheet to CSV. Everything is gated server-side too — this page just draws it. */
window.pageInit = async (S) => {
  const view = S.view();
  if (S.user.role !== "super_admin") {
    view.innerHTML = `<div class="empty card pad" style="margin-top:30px">Payroll is restricted to Super Admins.</div>`;
    return;
  }

  const peso = (n) => "₱" + Number(n || 0).toLocaleString("en-PH", { minimumFractionDigits: 2, maximumFractionDigits: 2 });
  const nowPeriod = new Date().toLocaleDateString("en-CA", { timeZone: "Asia/Manila" }).slice(0, 7);
  const monthLabel = (p) => { const [y, m] = p.split("-"); return new Date(y, m - 1, 1).toLocaleDateString("en-PH", { month: "long", year: "numeric" }); };
  let period = nowPeriod;
  let data = null;

  view.innerHTML = `
    <div class="pagehead"><div>
      <h2>Payroll &amp; Accounting</h2>
      <div class="lead">Net pay computed from attendance, overtime and leave. Adjust salaries and bonuses, then finalize the month.</div>
    </div></div>

    <div class="row between wrap" style="gap:12px;margin-bottom:16px">
      <label class="field" style="max-width:220px;margin:0">
        <span>Pay period</span>
        <input type="month" id="pr-month" value="${period}" max="${nowPeriod}">
      </label>
      <div class="row" style="gap:8px;align-items:flex-end">
        <a class="btn ghost" id="pr-csv" download>${S.ICON.download}Export CSV</a>
      </div>
    </div>

    <div id="pr-stats" class="pr-stats" style="display:grid;grid-template-columns:repeat(auto-fit,minmax(150px,1fr));gap:12px;margin-bottom:18px"></div>
    <div id="pr-body"></div>`;

  // Scoped styling for the stat cards + net-pay emphasis (theme-aware via CSS vars).
  const style = document.createElement("style");
  style.textContent = `
    .pr-stat{background:var(--card);border:1px solid var(--line);border-radius:16px;padding:14px 16px}
    .pr-stat .k{font-size:12px;color:var(--muted);font-weight:600;letter-spacing:.02em}
    .pr-stat .v{font-size:20px;font-weight:800;margin-top:4px;letter-spacing:-.01em}
    .pr-stat.net{background:var(--grad-green);border:none;color:#fff;box-shadow:var(--glow-green)}
    .pr-stat.net .k,.pr-stat.net .v{color:#fff}
    td.num,th.num{text-align:right;font-variant-numeric:tabular-nums;white-space:nowrap}
    .pr-net{font-weight:800}
    .lock{opacity:.6}`;
  view.appendChild(style);

  const monthEl = S.qs("#pr-month");
  monthEl.onchange = () => { period = monthEl.value || nowPeriod; load(); };

  function syncCsv() {
    const a = S.qs("#pr-csv");
    a.href = `/api/payroll?period=${encodeURIComponent(period)}&export=csv`;
  }

  function renderStats(t) {
    S.qs("#pr-stats").innerHTML = `
      <div class="pr-stat"><div class="k">Base salaries</div><div class="v">${peso(t.base)}</div></div>
      <div class="pr-stat"><div class="k">Overtime pay</div><div class="v">${peso(t.overtime)}</div></div>
      <div class="pr-stat"><div class="k">Bonuses</div><div class="v">${peso(t.bonus)}</div></div>
      <div class="pr-stat"><div class="k">Deductions</div><div class="v">${peso(t.deductions)}</div></div>
      <div class="pr-stat net"><div class="k">Net payout — ${S.esc(monthLabel(period))}</div><div class="v">${peso(t.net)}</div></div>`;
  }

  function renderTable(rows, workingDays) {
    const body = S.qs("#pr-body");
    if (!rows.length) { body.innerHTML = `<div class="empty">No active employees to pay.</div>`; return; }
    body.innerHTML = `
      <div class="lead" style="margin-bottom:8px">${workingDays} working days this month · overtime paid at 1.25× · unpaid leave and absences are deducted at the daily rate.</div>
      <div class="table-wrap"><table>
        <thead><tr>
          <th>Employee</th>
          <th class="num">Monthly salary</th>
          <th class="num">Present</th>
          <th class="num">Absent</th>
          <th class="num">Unpaid</th>
          <th class="num">OT (h)</th>
          <th class="num">OT pay</th>
          <th class="num">Bonus</th>
          <th class="num">Deductions</th>
          <th class="num">Net pay</th>
          <th style="text-align:right">Actions</th>
        </tr></thead>
        <tbody>${rows.map((r) => {
          const ded = (r.deduction || 0) + (r.absence_deduction || 0);
          return `<tr class="${r.finalized ? "lock" : ""}">
            <td>${S.avatar(r)}<div style="display:inline-block;vertical-align:middle;margin-left:8px">
              <div style="font-weight:600">${S.esc(r.name)}${r.finalized ? ' <span class="pill green" title="Finalized — locked">Final</span>' : ""}</div>
              <div class="sub" style="font-size:12px;color:var(--muted)">${S.esc(r.email || "")}</div></div></td>
            <td class="num">${r.monthly_salary ? peso(r.monthly_salary) : '<span style="color:var(--muted)">—</span>'}</td>
            <td class="num">${r.present_days}</td>
            <td class="num">${r.absent_days || 0}</td>
            <td class="num">${r.unpaid_leave_days || 0}</td>
            <td class="num">${r.overtime_hours || 0}</td>
            <td class="num">${peso(r.overtime_pay)}</td>
            <td class="num">${peso(r.bonus)}</td>
            <td class="num">${peso(ded)}</td>
            <td class="num pr-net">${peso(r.net_pay)}</td>
            <td style="text-align:right;white-space:nowrap">
              <button class="btn sm ghost" data-edit="${r.user_id}">Edit</button>
            </td></tr>`;
        }).join("")}</tbody></table></div>`;
    S.qsa("[data-edit]").forEach((b) => b.onclick = () => openEdit(rows.find((r) => r.user_id == b.dataset.edit)));
  }

  function openEdit(r) {
    const m = S.modal({
      title: `${r.name} — ${monthLabel(period)}`,
      body: `
        <label class="field"><span>Monthly salary (₱)</span>
          <input type="number" id="pr-salary" min="0" step="0.01" value="${r.monthly_salary || 0}"></label>
        <div class="row" style="gap:10px">
          <label class="field" style="flex:1"><span>Bonus (₱)</span>
            <input type="number" id="pr-bonus" min="0" step="0.01" value="${r.bonus || 0}"></label>
          <label class="field" style="flex:1"><span>Deduction (₱)</span>
            <input type="number" id="pr-ded" min="0" step="0.01" value="${r.deduction || 0}"></label>
        </div>
        <label class="field"><span>Note (optional)</span>
          <input type="text" id="pr-note" value="${S.esc(r.note || "")}" placeholder="e.g. 13th month advance, cash advance"></label>
        <div class="card pad" style="background:var(--canvas);font-size:13px">
          <div class="row between"><span>Auto deduction (absences + unpaid leave)</span><strong>${peso(r.absence_deduction)}</strong></div>
          <div class="row between"><span>Overtime pay (${r.overtime_hours || 0}h)</span><strong>${peso(r.overtime_pay)}</strong></div>
        </div>
        <label class="chip" style="cursor:pointer;margin-top:12px;display:inline-flex;align-items:center;gap:8px">
          <input type="checkbox" id="pr-final" style="width:auto" ${r.finalized ? "checked" : ""}> Mark this month finalized (locks the row)</label>`,
      footer: `<button class="btn ghost" id="pr-cancel">Cancel</button><button class="btn primary" id="pr-save">Save</button>`,
    });
    S.qs("#pr-cancel").onclick = m.close;
    S.qs("#pr-save").onclick = async () => {
      const salary = Number(S.qs("#pr-salary").value || 0);
      const bonus = Number(S.qs("#pr-bonus").value || 0);
      const deduction = Number(S.qs("#pr-ded").value || 0);
      const note = S.qs("#pr-note").value.trim() || null;
      const finalized = S.qs("#pr-final").checked;
      try {
        // Unlock first so the adjust write is never blocked by a finalized period.
        if (r.finalized) await S.api(`/api/payroll/finalize/${r.user_id}`, { method: "POST", body: { period, finalized: false } });
        if (salary !== (r.monthly_salary || 0)) await S.api(`/api/payroll/salary/${r.user_id}`, { method: "PUT", body: { monthly_salary: salary } });
        await S.api(`/api/payroll/adjust/${r.user_id}`, { method: "POST", body: { period, bonus, deduction, note } });
        if (finalized) await S.api(`/api/payroll/finalize/${r.user_id}`, { method: "POST", body: { period, finalized: true } });
        S.toast("Payroll updated", "ok");
        m.close(); load();
      } catch (e) { S.toast(e.detail || "Couldn't save", "err"); }
    };
  }

  async function load() {
    syncCsv();
    const body = S.qs("#pr-body");
    body.innerHTML = '<div class="skeleton" style="height:200px"></div>';
    try {
      data = await S.api(`/api/payroll?period=${encodeURIComponent(period)}`);
    } catch (e) {
      body.innerHTML = `<div class="empty">${S.esc(e.detail || "Failed to load payroll")}</div>`;
      return;
    }
    renderStats(data.totals);
    renderTable(data.rows, data.working_days);
  }

  load();
};

File: backend/app/routers/payroll.py
"""Payroll / Accounting — Super Admin ONLY. Computes each employee's net pay for a month from the
attendance, overtime, and leave data already in Sentinel, with editable per-person bonus/deduction.

Every route is gated by ``require_roles(ROLE_SUPER_ADMIN)`` so salaries are never exposed to other
roles — not just hidden in the UI. All salary/bonus changes are written to the audit trail.
"""
from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import select
from sqlalchemy.orm import Session

from ..constants import ROLE_SUPER_ADMIN
from ..database import get_db
from ..models import PayrollEntry, User
from ..schemas import PayrollAdjustIn, PayrollFinalizeIn, SalaryIn
from ..security import require_roles
from ..services import audit
from ..services import payroll as payroll_svc
from ..utils.csv_export import csv_response

router = APIRouter(prefix="/api/payroll", tags=["payroll"])

_SA = require_roles(ROLE_SUPER_ADMIN)


def _valid_period(period: str) -> str:
    try:
        y, m = int(period[:4]), int(period[5:7])
        date(y, m, 1)
        if period[4] != "-" or not (1 <= m <= 12):
            raise ValueError
    except (ValueError, IndexError):
        raise HTTPException(status_code=422, detail="period must be YYYY-MM")
    return f"{y:04d}-{m:02d}"


@router.get("")
def get_payroll(
    period: str = Query(...),
    export: str | None = Query(None),
    _: User = Depends(_SA),
    db: Session = Depends(get_db),
):
    period = _valid_period(period)
    result = payroll_svc.compute_period(db, period)
    if export == "csv":
        headers = ["Employee", "Email", "Monthly Salary", "Present", "Absent",
                   "Unpaid Leave", "OT Hours", "OT Pay", "Bonus", "Deduction", "Net Pay", "Finalized"]
        rows = [
            [r["name"], r["email"], r["monthly_salary"], r["present_days"], r["absent_days"],
             r["unpaid_leave_days"], r["overtime_hours"], r["overtime_pay"], r["bonus"],
             r["deduction"] + r["absence_deduction"], r["net_pay"], "Yes" if r["finalized"] else "No"]
            for r in result["rows"]
        ]
        return csv_response(f"sentinel-payroll-{period}.csv", headers, rows)
    return result


@router.put("/salary/{user_id}")
def set_salary(
    user_id: int,
    payload: SalaryIn,
    admin: User = Depends(_SA),
    db: Session = Depends(get_db),
):
    user = db.get(User, user_id)
    if not user:
        raise HTTPException(status_code=404, detail="User not found")
    old = user.monthly_salary
    user.monthly_salary = payload.monthly_salary
    audit.record(db, actor_id=admin.id, table_name="users", record_id=user_id, action="set_salary",
                 old={"monthly_salary": old}, new={"monthly_salary": payload.monthly_salary}, commit=False)
    db.commit()
    return {"ok": True, "user_id": user_id, "monthly_salary": user.monthly_salary}


def _get_entry(db: Session, user_id: int, period: str) -> PayrollEntry | None:
    return db.execute(
        select(PayrollEntry).where(PayrollEntry.user_id == user_id, PayrollEntry.period == period)
    ).scalar_one_or_none()


@router.post("/adjust/{user_id}")
def adjust(
    user_id: int,
    payload: PayrollAdjustIn,
    admin: User = Depends(_SA),
    db: Session = Depends(get_db),
):
    if not db.get(User, user_id):
        raise HTTPException(status_code=404, detail="User not found")
    period = _valid_period(payload.period)
    entry = _get_entry(db, user_id, period)
    old = {"bonus": entry.bonus, "deduction": entry.deduction, "note": entry.note} if entry else None
    if entry:
        if entry.finalized:
            raise HTTPException(status_code=409, detail="Period is finalized - unlock it first")
        entry.bonus = payload.bonus
        entry.deduction = payload.deduction
        entry.note = payload.note
    else:
        entry = PayrollEntry(user_id=user_id, period=period, bonus=payload.bonus,
                             deduction=payload.deduction, note=payload.note)
        db.add(entry)
    audit.record(db, actor_id=admin.id, table_name="payroll_entries", record_id=f"{user_id}:{period}",
                 action="adjust", old=old,
                 new={"bonus": payload.bonus, "deduction": payload.deduction, "note": payload.note},
                 commit=False)
    db.commit()
    return {"ok": True}


@router.post("/finalize/{user_id}")
def finalize(
    user_id: int,
    payload: PayrollFinalizeIn,
    admin: User = Depends(_SA),
    db: Session = Depends(get_db),
):
    if not db.get(User, user_id):
        raise HTTPException(status_code=404, detail="User not found")
    period = _valid_period(payload.period)
    entry = _get_entry(db, user_id, period)
    if not entry:
        entry = PayrollEntry(user_id=user_id, period=period)
        db.add(entry)
    entry.finalized = payload.finalized
    audit.record(db, actor_id=admin.id, table_name="payroll_entries", record_id=f"{user_id}:{period}",
                 action="finalize", new={"finalized": payload.finalized}, commit=False)
    db.commit()
    return {"ok": True, "finalized": entry.finalized}
